engine/src/main/java/io/camunda/zeebe/engine/state/analyzers/CatchEventAnalyzer.java

```diff
@@ -135,13 +146,103 @@ private Either<List<DirectBuffer>, CatchEventTuple> findCatchEventInScope(
   }
 
   public Optional<CatchEventTuple> findEscalationCatchEvent(
-      final DirectBuffer escalationCode, final ElementInstance instance) {
-    // TODO walk through the scope hierarchy and look for a matching catch event
+      final DirectBuffer escalationCode, ElementInstance instance) {
+    // walk through the scope hierarchy and look for a matching catch event
+    while (instance != null && instance.isActive()) {
+      final var instanceRecord = instance.getValue();
+      final var process = getProcess(instanceRecord.getProcessDefinitionKey());
+
+      final var found = findEscalationCatchEventInProcess(escalationCode, process, instance);
+      if (found.isPresent()) {
+        return found;
+      }
+
+      // find in parent process instance if exists
+      final var parentElementInstanceKey = instanceRecord.getParentElementInstanceKey();
+      instance = elementInstanceState.getInstance(parentElementInstanceKey);
+    }
+
+    // no matching catch event found
+    return Optional.empty();
+  }
+
+  private Optional<CatchEventTuple> findEscalationCatchEventInProcess(
+      final DirectBuffer escalationCode,
+      final ExecutableProcess process,
+      ElementInstance instance) {
+
+    while (instance != null && instance.isActive() && !instance.isInterrupted()) {
+      final var found = findEscalationCatchEventInScope(escalationCode, process, instance);
+      if (found.isPresent()) {
+        return found;
+      }
+      // find in parent scope if exists
+      final var instanceParentKey = instance.getParentKey();
+      instance = elementInstanceState.getInstance(instanceParentKey);
```

Comment:
🙃 Again, not necessary, but this could be achieved with a recursive call as well 😄